src/Components/Web.Extensions/src/InputFile/RemoteBrowserFileStream.cs

```diff
@@ -0,0 +1,156 @@
+            RemoteBrowserFileStreamOptions options,
+            CancellationToken cancellationToken)
+            : base(file)
+        {
+            _jsRuntime = jsRuntime;
+            _inputFileElement = inputFileElement;
+            _maxSegmentSize = options.SegmentSize;
+            _segmentFetchTimeout = options.SegmentFetchTimeout;
+
+            var pipe = new Pipe(new PipeOptions(pauseWriterThreshold: options.MaxBufferSize, resumeWriterThreshold: options.MaxBufferSize));
+            _pipeReader = pipe.Reader;
+            _fillBufferCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+
+            _ = FillBuffer(pipe.Writer, _fillBufferCts.Token);
+        }
+
+        private async Task FillBuffer(PipeWriter writer, CancellationToken cancellationToken)
+        {
+            long offset = 0;
+
+            while (offset < File.Size)
+            {
+                var pipeBuffer = writer.GetMemory(_maxSegmentSize);
+                var segmentSize = (int)Math.Min(_maxSegmentSize, File.Size - offset);
+
+                try
+                {
+                    using var readSegmentCts = CancellationTokenSource.CreateLinkedTokenSource(_fillBufferCts.Token);
+                    readSegmentCts.CancelAfter(_segmentFetchTimeout);
+
+                    var bytes = await _jsRuntime.InvokeAsync<byte[]>(
+                        InputFileInterop.ReadFileData,
+                        readSegmentCts.Token,
+                        _inputFileElement,
+                        File.Id,
+                        offset,
+                        segmentSize);
+
+                    if (bytes.Length != segmentSize)
+                    {
+                        throw new InvalidOperationException(
+                            $"A segment with size {bytes.Length} bytes was received, but {segmentSize} bytes were expected.");
+                    }
+
+                    bytes.CopyTo(pipeBuffer);
+                    writer.Advance(segmentSize);
+                    offset += segmentSize;
+
+                    var result = await writer.FlushAsync(cancellationToken);
+
+                    if (result.IsCompleted)
+                    {
+                        break;
+                    }
+                }
+                catch (Exception e)
+                {
+                    await writer.CompleteAsync(e);
+                    return;
+                }
+            }
+
+            await writer.CompleteAsync();
+        }
+
+        protected override async ValueTask<int> CopyFileDataIntoBuffer(long sourceOffset, Memory<byte> destination, CancellationToken cancellationToken)
+        {
+            if (_isReadingCompleted)
+            {
+                return 0;
+            }
+
+            int totalBytesCopied = 0;
+
+            while (destination.Length > 0)
+            {
+                var result = await _pipeReader.ReadAsync(cancellationToken);
+
+                if (result.IsCanceled)
```

Comment:
Since no one ever calls `_pipeReader.CancelPendingRead()`, `ReadResult.IsCanceled` can never be true, so this if block can be safely removed.